crates/wasi-runtime-config/src/lib.rs

```diff
@@ -0,0 +1,68 @@
+use anyhow::Result;
+use std::collections::HashMap;
+
+mod gen_ {
+    wasmtime::component::bindgen!({
+        path: "wit",
+        world: "wasi:config/imports",
+        trappable_imports: true,
+    });
+}
+use self::gen_::wasi::config::runtime as generated;
+
+/// Capture the state necessary for use in the `wasi-runtime-config` API implementation.
+pub struct WasiRuntimeConfigCtx {
+    runtime_config: HashMap<String, String>,
+}
+
+impl WasiRuntimeConfigCtx {
+    /// Create a new context.
+    pub fn new<S, I>(config: I) -> Self
+    where
+        S: Into<String>,
+        I: IntoIterator<Item = (S, S)>,
+    {
+        Self {
+            runtime_config: config
+                .into_iter()
+                .map(|(k, v)| (k.into(), v.into()))
+                .collect(),
+        }
+    }
+}
+
+/// A wrapper capturing the needed internal `wasi-runtime-config` state.
+pub struct WasiRuntimeConfigView<'a> {
+    ctx: &'a WasiRuntimeConfigCtx,
+}
+
+impl<'a> WasiRuntimeConfigView<'a> {
+    /// Create a new view into the `wasi-runtime-config` state.
+    pub fn new(ctx: &'a WasiRuntimeConfigCtx) -> Self {
+        Self { ctx }
+    }
+}
+
+impl generated::Host for WasiRuntimeConfigView<'_> {
+    fn get(&mut self, key: String) -> Result<Result<Option<String>, generated::ConfigError>> {
+        Ok(Ok(self.ctx.runtime_config.get(&key).map(|s| s.to_owned())))
+    }
+
+    fn get_all(&mut self) -> Result<Result<Vec<(String, String)>, generated::ConfigError>> {
+        Ok(Ok(self
+            .ctx
+            .runtime_config
+            .iter()
+            .map(|(k, v)| (k.to_string(), v.to_string()))
+            .collect()))
+    }
+}
+
+/// Add all the `wasi-runtime-config` world's interfaces to a [`wasmtime::component::Linker`].
+pub fn add_to_linker<T>(
+    l: &mut wasmtime::component::Linker<T>,
+    f: impl Fn(&mut T) -> WasiRuntimeConfigView + Send + Sync + Copy + 'static,
```

Comment:
Isn't this `... + 'static` bound going to mean that you can effectively only use `WasiRuntimeConfigView<'static>`?

If so -- and if the `'static` bound is necessary, which I think it is -- then instead of using lifetimes and borrows between `WasiRuntimeConfigVariables` and `WasiRuntimeConfig`, we should make `WasiRuntimeConfig` a newtype of `Arc<WasiRuntimeConfigVariables>`.

Reply:
The `'static` bound is necessary, but this made me think further, since the `wasi-runtime-config` API is very simple, so I reduced some excessive encapsulation by removing `WasiRuntimeConfigVariables`/`WasiRuntimeConfigCtx` and keeping only `WasiRuntimeConfig`. Now, this problem should no longer exists.

Reply:
The `+ 'static` doesn't require `<'static>` in this case since the `+ 'static` only applies to the lifetime of the captures of the closure itself. This I think was the desired signature which allows borrowing the config without cloning it on import calls.